lib/nss.c

```diff
@@ -0,0 +1,150 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <dlfcn.h>
+#include <stdbool.h>
+#include <string.h>
+#include <strings.h>
+#include <ctype.h>
+#include "prototypes.h"
+#include "../libsubid/subid.h"
+
+#define NSSWITCH "/etc/nsswitch.conf"
+
+// NSS plugin handling for subids
+// If nsswitch has a line like
+//    subid: sssd
+// then sssd will be consulted for subids.  Unlike normal NSS dbs,
+// only one db is supported at a time.  That's open to debate, but
+// the subids are a pretty limited resource, and local files seem
+// bound to step on any other allocations leading to insecure
+// conditions.
+static bool nss_initialized;
+
+static struct subid_nss_ops *subid_nss;
+
+struct subid_nss_ops *get_subid_nss_handle() {
+	return subid_nss;
+}
+
+bool nss_is_initialized() {
+	return nss_initialized;
+}
+
+void nss_exit() {
+	if (nss_initialized && subid_nss) {
+		dlclose(subid_nss->handle);
+		free(subid_nss);
+		subid_nss = NULL;
+	}
+}
+
+// nsswitch_path is an argument only to support testing.
+void nss_init(char *nsswitch_path) {
+	FILE *nssfp = NULL;
+	char *line = NULL, *p, *token;
+	size_t len = 0;
+
+	if (nss_initialized)
+		return;
+
+	if (!nsswitch_path)
+		nsswitch_path = NSSWITCH;
+
+	// read nsswitch.conf to check for a line like:
+	//   subid:	files
+	nssfp = fopen(nsswitch_path, "r");
+	if (!nssfp) {
+		fprintf(stderr, "Failed opening %s: %m", nsswitch_path);
+		nss_initialized = true; // let's not keep trying
+		return;
+	}
+	while ((getline(&line, &len, nssfp)) != -1) {
+		if (line[0] == '\0' || line[0] == '#')
+			continue;
+		if (strlen(line) < 8)
+			continue;
+		if (strncasecmp(line, "subid:", 6) != 0)
+			continue;
+		p = &line[6];
+		while ((*p) && isspace(*p))
+			p++;
+		if (!*p)
+			continue;
+		for (token = strtok(p, " \n\t");
```

Comment:
could we use `strtok_r` here to avoid surprises if another thread is also using `strtok`?

Or do you suggest using `nss_init(NULL)` before there are multiple threads running?

Reply:
Additionally, `nss_init()` is thread-unsafe in general, this might be an issue for `libsubid` users...

Reply:
Yeah - I'll go ahead and add locking and use _r.